source/math/Quaternion.ooc

```diff
@@ -91,7 +114,73 @@ Quaternion: cover {
 		else
 			roll = asin((upYawPitch scalarProduct(upRotated) * upYawPitch z - upRotated z) / rightYawPitch z)
 
-		FloatRotation3D new(pitch, -yaw, roll)
+		FloatRotation3D new (pitch, -yaw, roll)
+	}
+	distance: func (other: Quaternion) -> Float {
+		(this - other) norm
+	}
+	rotationX: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() < this precision)
+				result = 0.0f
+			else
+				result = (2.0f * (this w * this x + this y * this z)) atan2(1.0f - 2.0f * (this x squared() + this y squared()))
+			result
+		}
+	}
+	rotationY: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() > this precision)
+				result = ((2.0f * (this w * this y - this z * this x)) clamp(-1, 1)) asin()
+			else
+				result = Float sign(value) * (Float pi / 2.0f)
+			result
+		}
+	}
+	rotationZ: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() < this precision)
+				result = 2.0f * (this z atan2(this w))
+			else
+				result = (2.0f * (this w * this z + this x * this y)) atan2(1.0f - 2.0f * (this y squared() + this z squared()))
+			result
+		}
+	}
+	direction: FloatPoint3D {
+		get {
+			quaternionLogarithm := this logarithm
+			quaternionLogarithm imaginary / quaternionLogarithm imaginary norm
+		}
+	}
+	logarithm: Quaternion {
+		get {
+			result: Quaternion
```

Comment:
You should be able to replace every occurence of `Quaternion` (except the first one) with `This`, including constructor calls, argument types, local variables, and return types. The word `Quaternion` should only occur once in this entire file, unless I'm mistaken.